Undefined symbols when building against external PMIX 3.1.5

I am trying to build current `master` (`g0bcef04`) against a site-provided PMIX 3.1.5. `configure` complained about the old version:

```
*** Configuring PRRTE
checking if PMIx version is 4.0.0 or greater... no
configure: WARNING: OMPI's internal runtime environment "PRRTE" does not support
configure: WARNING: PMIx versions less than v4.x as they lack adequate tool
configure: WARNING: support. You can, if desired, build OMPI against an earlier
configure: WARNING: version of PMIx for strictly direct-launch purposes - e.g., using)
configure: WARNING: Slurm's srun to launch the job - by configuring with the
configure: WARNING: --disable-internal-rte option.
configure: error: Cannot continue
```

so I ran configure with `--disable-internal-rte`, which succeeds but then the build fails:

```
  CC       runtime/ompi_rte.lo
In file included from ../../opal/util/proc.h:25:0,
                 from ../../ompi/runtime/ompi_rte.c:46:
../../ompi/runtime/ompi_rte.c: In function 'ompi_rte_init':
../../ompi/runtime/ompi_rte.c:673:40: error: 'PMIX_APP_ARGV' undeclared (first use in this function); did you mean 'PMIX_APP_FREE'?
     OPAL_MODEX_RECV_VALUE_OPTIONAL(rc, PMIX_APP_ARGV,
                                        ^
../../opal/mca/pmix/pmix-internal.h:293:35: note: in definition of macro 'OPAL_MODEX_RECV_VALUE_OPTIONAL'
         (r) = PMIx_Get(&(_proc), (s), &(_info), 1, &(_kv));                             \
                                   ^
../../ompi/runtime/ompi_rte.c:673:40: note: each undeclared identifier is reported only once for each function it appears in
     OPAL_MODEX_RECV_VALUE_OPTIONAL(rc, PMIX_APP_ARGV,
                                        ^
../../opal/mca/pmix/pmix-internal.h:293:35: note: in definition of macro 'OPAL_MODEX_RECV_VALUE_OPTIONAL'
         (r) = PMIx_Get(&(_proc), (s), &(_info), 1, &(_kv));                             \
                                   ^
../../ompi/runtime/ompi_rte.c:686:40: error: 'PMIX_REINCARNATION' undeclared (first use in this function); did you mean 'PMIX_INTERNAL'?
     OPAL_MODEX_RECV_VALUE_OPTIONAL(rc, PMIX_REINCARNATION,
                                        ^
../../opal/mca/pmix/pmix-internal.h:293:35: note: in definition of macro 'OPAL_MODEX_RECV_VALUE_OPTIONAL'
         (r) = PMIx_Get(&(_proc), (s), &(_info), 1, &(_kv));                             \
                                   ^
make[2]: *** [runtime/ompi_rte.lo] Error 1
```

Here is the output of `pmix_info`, in case that matters:

```
$ pmix_info --version
                 Package: PMIx root@tauruslogin6.taurus.hrsk.tu-dresden.de
                          Distribution
                    PMIX: 3.1.5
      PMIX repo revision: gitedebb24
       PMIX release date: Apr 27, 2020
                  Prefix: /usr
 Configured architecture: pmix.arch
          Configure host: tauruslogin6.taurus.hrsk.tu-dresden.de
           Configured by: root
           Configured on: 2020-04-27T11:24:26
          Configure host: tauruslogin6.taurus.hrsk.tu-dresden.de
  Configure command line: '--build=x86_64-redhat-linux-gnu'
                          '--host=x86_64-redhat-linux-gnu'
                          '--program-prefix=' '--disable-dependency-tracking'
                          '--prefix=/usr' '--exec-prefix=/usr'
                          '--bindir=/usr/bin' '--sbindir=/usr/sbin'
                          '--sysconfdir=/etc' '--datadir=/usr/share'
                          '--includedir=/usr/include' '--libdir=/usr/lib64'
                          '--libexecdir=/usr/libexec' '--localstatedir=/var'
                          '--sharedstatedir=/var/lib'
                          '--mandir=/usr/share/man'
                          '--infodir=/usr/share/info'
                Built by: root
                Built on: 2020-04-27T11:25:58
              Built host: tauruslogin6.taurus.hrsk.tu-dresden.de
              C compiler: gcc
     C compiler absolute: /usr/bin/gcc
  C compiler family name: GNU
      C compiler version: 4.8.5
  Internal debug support: no
              dl support: yes
     Symbol vis. support: yes
              MCA bfrops: v3 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
              MCA bfrops: v12 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
              MCA bfrops: v20 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
              MCA bfrops: v21 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA gds: ds12 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA gds: hash (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA gds: ds21 (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA pdl: pdlopen (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA pif: linux_ipv6 (MCA v2.1.0, API v2.0.0, Component
                          v3.1.5)
                 MCA pif: posix_ipv4 (MCA v2.1.0, API v2.0.0, Component
                          v3.1.5)
        MCA pinstalldirs: env (MCA v2.1.0, API v1.0.0, Component v3.1.5)
        MCA pinstalldirs: config (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA plog: syslog (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA plog: stdfd (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA plog: default (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA pnet: test (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA pnet: tcp (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA preg: native (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA psec: none (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                MCA psec: native (MCA v2.1.0, API v1.0.0, Component v3.1.5)
             MCA psensor: file (MCA v2.1.0, API v1.0.0, Component v3.1.5)
             MCA psensor: heartbeat (MCA v2.1.0, API v1.0.0, Component
                          v3.1.5)
              MCA pshmem: mmap (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA ptl: tcp (MCA v2.1.0, API v1.0.0, Component v3.1.5)
                 MCA ptl: usock (MCA v2.1.0, API v1.0.0, Component v3.1.5)
```

Is this expected and is there another way to build Open MPI on such a system?

--- a/config/opal_config_pmix.m4
+++ b/config/opal_config_pmix.m4
@@ -13,7 +13,7 @@ dnl                         All rights reserved.
 dnl Copyright (c) 2009-2019 Cisco Systems, Inc.  All rights reserved.
 dnl Copyright (c) 2011-2014 Los Alamos National Security, LLC. All rights
 dnl                         reserved.
-dnl Copyright (c) 2014-2019 Intel, Inc.  All rights reserved.
+dnl Copyright (c) 2014-2020 Intel, Inc.  All rights reserved.
 dnl Copyright (c) 2014-2018 Research Organization for Information Science
 dnl                         and Technology (RIST).  All rights reserved.
 dnl Copyright (c) 2016      IBM Corporation.  All rights reserved.
@@ -166,11 +166,11 @@ AC_DEFUN([_OPAL_CONFIG_PMIX_EXTERNAL], [
            LIBS="$opal_pmix_LIBS_save $opal_pmix_LIBS"
 
            AS_IF([test "$opal_pmix_external_support" = "yes"],
-                 [AC_MSG_CHECKING([if external PMIx version is 3.0.0 or greater])
+                 [AC_MSG_CHECKING([if external PMIx version is 3.1.5 or greater])
                     AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <pmix_version.h>]],
                           [[
-#if PMIX_VERSION_MAJOR < 3L
-#error "pmix API version is less than 3.0.0"
+#if PMIX_NUMERIC_VERSION < 0x00030105
+#error "pmix API version is less than 3.1.5"
 #endif
                           ]])],
                           [AC_MSG_RESULT([yes])],

--- a/ompi/dpm/dpm.c
+++ b/ompi/dpm/dpm.c
@@ -696,7 +696,11 @@ static int dpm_convert(opal_list_t *infos,
                 /* does it conflict? */
                if (0 != strncasecmp(ck, directive, strlen(directive))) {
                     opal_asprintf(&help_str, "Conflicting directives \"%s %s\"", ck, directive);
+#if PMIX_NUMERIC_VERSION >= 0x00040000
                     attr = PMIx_Get_attribute_string(option);
+#else
+                    attr = option;
+#endif
                     opal_show_help("help-dpm.txt", "deprecated-fail", true,
                                    infokey, attr, help_str);
                     free(help_str);
@@ -723,7 +727,11 @@ static int dpm_convert(opal_list_t *infos,
                     if (OMPI_SUCCESS != rc) {
                         /* we have a conflict */
                         opal_asprintf(&ptr, "  Option %s\n  Conflicting modifiers \"%s %s\"", option, infokey, modifier);
+#if PMIX_NUMERIC_VERSION >= 0x00040000
                         attr = PMIx_Get_attribute_string(option);
+#else
+                        attr = option;
+#endif
                         opal_show_help("help-dpm.txt", "deprecated-fail", true,
                                        infokey, attr, ptr);
                         free(ptr);
@@ -983,9 +991,9 @@ int ompi_dpm_spawn(int count, const char *array_of_commands[],
 
             /* non-standard keys
              * Keys that correspond to prun/mpiexec parameters
-             * do not deprecate PMIX unprefixed forms to remain identical 
+             * do not deprecate PMIX unprefixed forms to remain identical
              * to the command line parameter;
-             * Keys that are not corresponding to an mpiexec parameter are 
+             * Keys that are not corresponding to an mpiexec parameter are
              * deprecated in the non-prefixed form */
 
             /* check for 'hostfile' */

--- a/ompi/runtime/ompi_rte.c
+++ b/ompi/runtime/ompi_rte.c
@@ -668,6 +668,7 @@ int ompi_rte_init(int *pargc, char ***pargv)
         }
     }
 
+#ifdef PMIX_APP_ARGV
     /* get our command - defaults to our appnum */
     ev1 = NULL;
     OPAL_MODEX_RECV_VALUE_OPTIONAL(rc, PMIX_APP_ARGV,
@@ -681,13 +682,21 @@ int ompi_rte_init(int *pargc, char ***pargv)
             opal_process_info.command = opal_argv_join(tmp, ' ');
         }
     }
+#else
+    tmp = *pargv;
+    if (NULL != tmp) {
+        opal_process_info.command = opal_argv_join(tmp, ' ');
+    }
+#endif
 
+#ifdef PMIX_REINCARNATION
     /* get our reincarnation number */
     OPAL_MODEX_RECV_VALUE_OPTIONAL(rc, PMIX_REINCARNATION,
                                    &OPAL_PROC_MY_NAME, &u32ptr, PMIX_UINT32);
     if (PMIX_SUCCESS == rc) {
         opal_process_info.reincarnation = u32;
     }
+#endif
 
     /* get the number of local peers - required for wireup of
      * shared memory BTL, defaults to local node */

--- a/ompi/runtime/ompi_rte.c
+++ b/ompi/runtime/ompi_rte.c
@@ -585,9 +585,11 @@ int ompi_rte_init(int *pargc, char ***pargv)
             /* just assume 0 */
             u16 = 0;
         } else {
-            ret = opal_pmix_convert_status(rc);
-            error = "node rank";
-            goto error;
+            /* we may be in an environment that doesn't quite adhere
+             * to the Standard - we can safely assume it is the same
+             * as the local rank as such environments probably aren't
+             * going to care */
+            u16 = opal_process_info.my_local_rank;
         }
     }
     opal_process_info.my_node_rank = u16;
@@ -704,10 +706,6 @@ int ompi_rte_init(int *pargc, char ***pargv)
                                    &pname, &u32ptr, PMIX_UINT32);
     if (PMIX_SUCCESS == rc) {
         opal_process_info.num_local_peers = u32 - 1;  // want number besides ourselves
-    } else {
-        ret = opal_pmix_convert_status(rc);
-        error = "local size";
-        goto error;
     }
 
     /* retrieve temp directories info */
@@ -777,28 +775,27 @@ int ompi_rte_init(int *pargc, char ***pargv)
         opal_process_info.proc_is_bound = false;
     }
 
-    /* get our local peers */
-    if (0 < opal_process_info.num_local_peers) {
-        /* if my local rank if too high, then that's an error */
-        if (opal_process_info.num_local_peers < opal_process_info.my_local_rank) {
-            ret = OPAL_ERR_BAD_PARAM;
-            error = "num local peers";
-            goto error;
-        }
-        /* retrieve the local peers - defaults to local node */
-        val = NULL;
-        OPAL_MODEX_RECV_VALUE(rc, PMIX_LOCAL_PEERS,
-                              &pname, &val, PMIX_STRING);
-        if (PMIX_SUCCESS == rc && NULL != val) {
-            peers = opal_argv_split(val, ',');
-            free(val);
-        } else {
-            ret = opal_pmix_convert_status(rc);
-            error = "local peers";
-            goto error;
-        }
+    /* retrieve the local peers - defaults to local node */
+    val = NULL;
+    OPAL_MODEX_RECV_VALUE(rc, PMIX_LOCAL_PEERS,
+                          &pname, &val, PMIX_STRING);
+    if (PMIX_SUCCESS == rc && NULL != val) {
+        peers = opal_argv_split(val, ',');
+        free(val);
     } else {
-        peers = NULL;
+        ret = opal_pmix_convert_status(rc);
+        error = "local peers";
+        goto error;
+    }
+    /* if we were unable to retrieve the #local peers, set it here */
+    if (0 == opal_process_info.num_local_peers) {
+        opal_process_info.num_local_peers = opal_argv_count(peers) - 1;
+    }
+    /* if my local rank if too high, then that's an error */
+    if (opal_process_info.num_local_peers < opal_process_info.my_local_rank) {
+        ret = OPAL_ERR_BAD_PARAM;
+        error = "num local peers";
+        goto error;
     }
 
     /* set the locality */

